README.md
# flutter_unipay_plugin

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

根据[银联手机控件支付](https://open.unionpay.com/tjweb/acproduct/list?apiservId=450)的集成需要，为fluter平台开发一款能够实现基本功能的插件

## 使用说明

- 请参考[Flutter文档](https://flutter.io/)，了解plugin的一般用法
- 本插件在iOS platform集成时，由于银联静态库存在C++/ObjC混编，因此需要在你的Flutter工程内使用到银联API的地方，使用ObjectC++混编，具体请参考[example](https://github.com/delapecci/flutter_unipay_plugin/tree/master/example)里的ios工程配置

## 欢迎支持
- MIT协议，欢迎PR
- 具体开发请参考[Flutter插件开发文档](https://flutter.io/developing-packages/#edit-plugin-package)


CHANGELOG.md
## 0.0.1

* 2018-11-30 银联全渠道产品-手机控件支付集成的flutter插件，能够正常发起支付，并通过BinaryMessage向插件调用一端返回结果 


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_unipay_plugin_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(new MyApp());

    // Verify that platform version is retrieved.
    expect(
        find.byWidgetPredicate(
          (Widget widget) =>
              widget is Text && widget.data.startsWith('Running on:'),
        ),
        findsOneWidget);
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_unipay_plugin/flutter_unipay_plugin.dart';

import 'package:dio/dio.dart';
import 'dart:convert';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';
  String _tn = '';
  String _payResult = '';

  @override
  void initState() {
    super.initState();
    initPlatformState();
    getTestTn();
    const channel = BasicMessageChannel<String>(FlutterUnipayPlugin.MESSAGE_CHANNEL_NAME,
        StringCodec());
    channel.setMessageHandler((String message) async {
//      Map<String, String> payResult = jsonDecode(message); // 解析返回报文(json)格式
      print('支付结果: $message');
      setState(() {
        _payResult = message;
      });
      return '已收到支付结果';
    });
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String platformVersion;
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      platformVersion = await FlutterUnipayPlugin.platformVersion;
    } on PlatformException {
      platformVersion = 'Failed to get platform version.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = platformVersion;
    });
  }

  Future<void> getTestTn() async {
    Dio dio = new Dio();
    dio.options.connectTimeout = 120000;
    Response response = await dio.get("http://101.231.204.84:8091/sim/getacptn");
    if (!mounted) return;

    setState(() {
      _tn = response.data;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: new Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text('Running on: $_platformVersion\n'),
              Text('Tx Seq No: $_tn\n'),
              Text('Pay Result: $_payResult\n'),
              new RaisedButton(
                onPressed: () {
                  FlutterUnipayPlugin.upPay(this._tn, "01");
                },
                child: const Text('Pay'),
              )
            ]
        )
      ),
    );
  }
}


lib/flutter_unipay_plugin.dart
import 'dart:async';

import 'package:flutter/services.dart';

/// 银联全渠道-手机控件支付 flutter插件
/// 
class FlutterUnipayPlugin {

  /// 方法调用信道名称
  static const String METHOD_CHANNEL_NAME = 'com.nbp.flutter_unipay_plugin';
  /// 简单字符串消息传递信道名称
  static const String MESSAGE_CHANNEL_NAME = 'com.nbp.msg.flutter_unipay_plugin';
  /// 支付结果: 正常
  static const String UP_PAY_SUCCESS = "0000";
  /// 支付结果: 失败
  static const String UP_PAY_FAIL = "9999";
  /// 支付结果: 异常发生
  static const String UP_PAY_EXCEPTION = "8888";
  /// 支付结果: 支付完成无结果
  static const String UP_PAY_DONE = "6666";
  /// 支付结果: 支付取消
  static const String UP_PAY_CANCEL = "7777";
  /// 支付结果报文json格式 { 'code':'' [, 'data':'', 'sign':'', 'error':''] }
  static const String UP_PAY_RESULT_K_CODE = 'code';
  /// 支付结果报文json格式 { 'code':'' [, 'data':'', 'sign':'', 'error':''] }
  static const String UP_PAY_RESULT_K_DATA = 'data';
  /// 支付结果报文json格式 { 'code':'' [, 'data':'', 'sign':'', 'error':''] }
  static const String UP_PAY_RESULT_K_SIGN = 'sign';
  /// 支付结果报文json格式 { 'code':'' [, 'data':'', 'sign':'', 'error':''] }
  static const String UP_PAY_RESULT_K_ERROR = 'error';

  /// 支付环境: "01" - 连接银联测试环境
  static const String UP_PAY_MODE_TEST = "01";
  /// 支付环境: "00" - 启动银联正式环境
  static const String UP_PAY_MODE_PROD = "00";

  static const MethodChannel _channel =
      const MethodChannel(METHOD_CHANNEL_NAME);

  static Future<String> get platformVersion async {
    final String version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  /// 发送交易请求流水号[tn]，接入[mode]对应的银联环境，进行支付
  /// 
  /// ```dart
  /// // 调用插件支付
  /// upPay("xxxxxxxxxxx", FlutterUnipayPlugin.UP_PAY_MODE_TEST);
  /// 
  /// // 监听支付结果
  /// const channel = BasicMessageChannel<String>(FlutterUnipayPlugin.MESSAGE_CHANNEL_NAME,
  ///   StringCodec());
  /// channel.setMessageHandler((String message) async {
  ///   print('支付结果: $message');
  ///   setState(() {
  ///     _payResult = message;
  ///   });
  ///   return '已收到支付结果';
  /// });
  /// ```
  static Future upPay(String tn, String mode) async {
    return await _channel.invokeMethod("upPay", {"up_pay_tn": tn, "up_pay_mode": mode});
  }
}


android/src/main/java/com/nbp/flutterunipayplugin/FlutterUnipayPlugin.java
package com.nbp.flutterunipayplugin;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.util.Log;

import com.unionpay.UPPayAssistEx;

import io.flutter.plugin.common.BasicMessageChannel;
import io.flutter.plugin.common.JSONUtil;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.plugin.common.StringCodec;

import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONStringer;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * FlutterUnipayPlugin
 */
public class FlutterUnipayPlugin implements MethodCallHandler, PluginRegistry.ActivityResultListener {

    //***************************************************************************************//
    // Plugin Interface Contract for all clients
    //
    //***************************************************************************************//
    public static final String CHANNEL_NAME = "com.nbp.flutter_unipay_plugin";
    public static final String MSG_CHANNEL_NAME = "com.nbp.msg.flutter_unipay_plugin";
    public static final String UP_PAY_SUCCESS = "0000";
    public static final String UP_PAY_FAIL = "9999";
    public static final String UP_PAY_EXCEPTION = "8888";
    public static final String UP_PAY_DONE = "6666";
    public static final String UP_PAY_CANCEL = "7777";



    private static final String LOG_TAG = "FlutterUnipayPlugin";
    private static final int PLUGIN_VALID = 0;
    private static final int PLUGIN_NOT_INSTALLED = -1;
    private static final int PLUGIN_NEED_UPGRADE = 2;

    private Activity activity;
    static MethodChannel channel;
    static BasicMessageChannel<String> stringMsgChannel;

    /*****************************************************************
     * mMode参数解释： "00" - 启动银联正式环境 "01" - 连接银联测试环境
     *****************************************************************/
    private String mMode = "01";

    private FlutterUnipayPlugin(Activity activity) {
        this.activity = activity;
    }

    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        channel = new MethodChannel(registrar.messenger(), CHANNEL_NAME);
        stringMsgChannel = new BasicMessageChannel<>(
                registrar.view(), MSG_CHANNEL_NAME, StringCodec.INSTANCE);
        final FlutterUnipayPlugin instance = new FlutterUnipayPlugin(registrar.activity());
        registrar.addActivityResultListener(instance);
        channel.setMethodCallHandler(instance);
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {

        switch (call.method) {
            case "getPlatformVersion":
                result.success("Android " + android.os.Build.VERSION.RELEASE);
                break;
            case "upPay":
                unPay(call, result);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void unPay(MethodCall call, MethodChannel.Result result) {
        // With given arguments in call, start up Unipay prepare
        final String tn = call.argument("up_pay_tn");
        // mMode参数解释：
        // 0 - 启动银联正式环境
        // 1 - 连接银联测试环境
        final String argMode = call.argument("up_pay_mode");
        if (argMode != null) this.mMode = argMode;
        int ret = UPPayAssistEx.startPay(this.activity, null, null, tn, this.mMode);
        if (ret == PLUGIN_NEED_UPGRADE || ret == PLUGIN_NOT_INSTALLED) {
            // 需要重新安装控件
            Log.e(LOG_TAG, " plugin not found or need upgrade!!!");

            AlertDialog.Builder builder = new AlertDialog.Builder(this.activity);
            builder.setTitle("提示");
            builder.setMessage("完成购买需要安装银联支付控件，是否安装？");

            final Context ctx = this.activity;

            builder.setNegativeButton("确定",
                    new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            UPPayAssistEx.installUPPayPlugin(ctx);
                            dialog.dismiss();
                        }
                    });

            builder.setPositiveButton("取消",
                    new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.dismiss();
                        }
                    });
            builder.create().show();

        }
    }

    @Override
    public boolean onActivityResult(int i, int i1, Intent data) {
        /*************************************************
         * 步骤3：处理银联手机支付控件返回的支付结果
         ************************************************/
        if (data == null) {
            return true;
        }

        Map<String, String> payload = new HashMap<>();
        /*
         * 支付控件返回字符串:success、fail、cancel 分别代表支付成功，支付失败，支付取消
         */
        String str = data.getExtras().getString("pay_result");
        if (str.equalsIgnoreCase("success")) {

            // 如果想对结果数据验签，可使用下面这段代码，但建议不验签，直接去商户后台查询交易结果
            // result_data结构见c）result_data参数说明
            if (data.hasExtra("result_data")) {
                String result = data.getExtras().getString("result_data");
                try {
                    JSONObject resultJson = new JSONObject(result);
                    String sign = resultJson.getString("sign");
                    String dataOrg = resultJson.getString("data");
                    // 此处的verify建议送去商户后台做验签
                    // 如要放在手机端验，则代码必须支持更新证书
                    boolean ret = verify(dataOrg, sign, this.mMode);
                    if (ret) {
                        // 验签成功，显示支付结果
                        payload.put("code", UP_PAY_SUCCESS);
                        payload.put("data", dataOrg );
                        payload.put("sign", sign);
                    } else {
                        // 验签失败
                        payload.put("code", UP_PAY_FAIL);
                    }
                } catch (JSONException e) {
                    payload.put("code", UP_PAY_EXCEPTION);
                }
            } else {
                // TODO: 结果result_data为成功时，去商户后台查询一下再展示成功
                payload.put("code", UP_PAY_DONE);
            }
        } else if (str.equalsIgnoreCase("fail")) {
            payload.put("code", UP_PAY_FAIL);
        } else if (str.equalsIgnoreCase("cancel")) {
            payload.put("code", UP_PAY_CANCEL);
        }
        final JSONObject payloadJson = new JSONObject(payload);
        stringMsgChannel.send(payloadJson.toString(), new BasicMessageChannel.Reply<String>() {
            @Override
            public void reply(String s) {
                Log.i(LOG_TAG, "收到Dart端响应" + String.valueOf(s));
            }
        });

        return true;
    }

    private boolean verify(String msg, String sign64, String mode) {
        // TODO: Send message signature to backend and verify it
        return true;
    }
}


ios/Classes/FlutterUnipayPlugin.m
#import "FlutterUnipayPlugin.h"
#import "UPPaymentControl.h"


@implementation FlutterUnipayPlugin 
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"com.nbp.flutter_unipay_plugin"
            binaryMessenger:[registrar messenger]];
    msgChannel = [FlutterBasicMessageChannel messageChannelWithName:@"com.nbp.msg.flutter_unipay_plugin" binaryMessenger:[registrar messenger] codec: [FlutterStringCodec sharedInstance]];
    UIViewController *viewController  = (UIViewController *)registrar.messenger;
    FlutterUnipayPlugin* instance = [[FlutterUnipayPlugin alloc] initWithViewController:viewController];
    
    [registrar addMethodCallDelegate:instance channel:channel];
    [registrar addApplicationDelegate:instance];
}

- (instancetype)initWithViewController:(UIViewController *)viewController {
    self = [super init];
    if (self) {
        self.viewController = viewController;
    }
    return self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else if ([@"upPay" isEqualToString:call.method]){
    [self startPayAction:call];
    result(nil);
  }else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)startPayAction:(FlutterMethodCall *)call {
    NSString *tnStr = call.arguments[@"up_pay_tn"];
    NSString *tnModeStr = call.arguments[@"up_pay_mode"];
    [[UPPaymentControl defaultControl] startPay:tnStr fromScheme:@"renpinhui" mode:tnModeStr viewController:self.viewController];
}

// UIApplicationDelegate 
- (BOOL) application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation {
    
    [[UPPaymentControl defaultControl] handlePaymentResult:url completeBlock:^(NSString *code, NSDictionary *data) {
        NSMutableDictionary *payload = [[NSMutableDictionary alloc] init];
        if([code isEqualToString:@"success"]) {
            if(data != nil){
                [payload setValue:@"0000" forKey:@"code"];
                [payload setValue:data[@"sign"] forKey:@"sign"];
                [payload setValue:data[@"data"] forKey:@"data"];
            } else {
                //结果code为成功时，去商户后台查询一下确保交易是成功的再展示成功
                [payload setValue:@"6666" forKey:@"code"];
            }
        }
        else if([code isEqualToString:@"fail"]) {
            //交易失败
            [payload setValue:@"9999" forKey:@"code"];
        }
        else if([code isEqualToString:@"cancel"]) {
            //交易取消
            [payload setValue:@"7777" forKey:@"code"];
        }
        NSData *payloadData = [NSJSONSerialization dataWithJSONObject:payload
                                                           options:0
                                                             error:nil];
        NSString *payloadMsg = [[NSString alloc] initWithData:payloadData encoding:NSUTF8StringEncoding];
        [msgChannel sendMessage:payloadMsg reply:^(id  _Nullable reply) {
            NSLog(@"%@", reply);
        }];
    }];
    
    return YES;
}

// NOTE: 9.0以后使用新API接口
- (BOOL)application:(UIApplication *)app openURL:(NSURL *)url options:(NSDictionary<NSString*, id> *)options
{
    [[UPPaymentControl defaultControl] handlePaymentResult:url completeBlock:^(NSString *code, NSDictionary *data) {
        
        NSMutableDictionary *payload = [[NSMutableDictionary alloc] init];
        if([code isEqualToString:@"success"]) {
            //如果想对结果数据验签，可使用下面这段代码，但建议不验签，直接去商户后台查询交易结果
            if(data != nil){
                [payload setValue:@"0000" forKey:@"code"];
                [payload setValue:data[@"sign"] forKey:@"sign"];
                [payload setValue:data[@"data"] forKey:@"data"];
            } else {
                //结果code为成功时，去商户后台查询一下确保交易是成功的再展示成功
                [payload setValue:@"6666" forKey:@"code"];
            }
        }
        else if([code isEqualToString:@"fail"]) {
            //交易失败
            [payload setValue:@"9999" forKey:@"code"];
        }
        else if([code isEqualToString:@"cancel"]) {
            //交易取消
            [payload setValue:@"7777" forKey:@"code"];
        }
        NSData *payloadData = [NSJSONSerialization dataWithJSONObject:payload
                                                              options:0
                                                                error:nil];
        NSString *payloadMsg = [[NSString alloc] initWithData:payloadData encoding:NSUTF8StringEncoding];
        [msgChannel sendMessage:payloadMsg reply:^(id  _Nullable reply) {
            NSLog(@"%@", reply);
        }];
    }];
    
    return YES;
}

@end
